Poromechanics Kernel added to TensorMechanics

WilkAndy opened the pull request:
Fixes #4792 

Poroelasticity is far from finished, but it is taking me so long to write this stuff, so I think it most efficient to let others people contribute, eg @pou036 , @tonkmr, @cpritam , @dschwen , @bwspenc and anyone else that can be "encouraged"!

In my opinion, the following needs to be done.  I WILL do these eventually, if no one else has time.
- [x] Doco on mooseframework.org.  I hope to start (and finish) this today.
- [x] Many more tests.  The tests i've added demonstrate the Kernel almost definitely does the right thing.  However, since poromechanics will be extensively used, many more tests are needed.  These are not just to test the C++, but also to demonstrate how to use poromechanics, and to (importantly) test convergence behaviour and accuracy.  There are lots of analytical solutions we can compare against.
- [x] Create a PoroMechanics Action.
- [x] Extend to multiphase situation.  This is fairly trivial, if a little ugly.  Then the 'scalar' is not a Variable, but a function of Variables determined at runtime, and the Jacobian entries are obviously a bit more complicated.  I toyed with the idea of the GradScalar Kernel having this built in from the start, but decided it'd be annoying for users.

WilkAndy committed:
Added PoroMechanicsCoupling Kernel.

Fixes #4792

moosebuild commented:
Results of testing 722acfb57deeb9ccc4f2e99e3f635d2f99232d4e using moose_PR_pre_check recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/13613

moosebuild commented:
Results of testing 50192943ce0ab14e9651f32d2a7ba607f8d77d4c using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/13617

moosebuild commented:
Results of testing 50192943ce0ab14e9651f32d2a7ba607f8d77d4c using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/13620

moosebuild commented:
Results of testing 50192943ce0ab14e9651f32d2a7ba607f8d77d4c using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/13618

moosebuild commented:
Results of testing 50192943ce0ab14e9651f32d2a7ba607f8d77d4c using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/13619

friedmud commented on modules/tensor_mechanics/src/kernels/GradScalar.C:
```diff
@@ -0,0 +1,53 @@
+Real
+GradScalar::computeQpJacobian()
+{
+  if (_var.number() != _scalar_var_num)
```
This can't be right... it will never be true because this is not a ScalarKernel

friedmud replied:
Also: For the same reason, I'm not a huge fan of the name as it seems to imply that it's _for_ Scalar variables... instead of just coupled to them.

Is there a better (maybe more physical) name you could use?

WilkAndy replied:
Aha, yes, i had forgotten that Scalar has a special meaning in MOOSE.  I meant to distinguish it from the Tensors encountered in TensorMechanics.  I'll change the name after you've checked the remainder.  Perhaps i'll just use GradVariable.

a

friedmud replied:
Try to use something more descriptive than `GradVariable`.  Remember that these Kernels show up by default in new applications...

WilkAndy committed:
Tests of PoroMechanicsCoupling Kernel.

Fixes #4792

moosebuild commented:
Results of testing ce3fe1273d4409b3659203ec5a7b5fb2312d494f using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/13660

WilkAndy commented:
I changed the name of the Kernel, as suggested by @friedmud ,

a

moosebuild commented:
Results of testing ce3fe1273d4409b3659203ec5a7b5fb2312d494f using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/13663

moosebuild commented:
Results of testing ce3fe1273d4409b3659203ec5a7b5fb2312d494f using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/13661

moosebuild commented:
Results of testing ce3fe1273d4409b3659203ec5a7b5fb2312d494f using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/13662

friedmud commented on modules/tensor_mechanics/src/kernels/PoroMechanicsCoupling.C:
```diff
@@ -0,0 +1,53 @@
+Real
+PoroMechanicsCoupling::computeQpJacobian()
+{
+  if (_var.number() != _porepressure_var_num)
```
This if statement still doesn't make sense to me.  How can it ever be true?

WilkAndy replied:
Usually _var.number != _porepressure_var_num, @friedmud .   The _var is usually disp_x (disp_y or disp_z), and the porepressure is another variable.  I included line 44 just in case porepressure = disp_x (or similar).

tonkmr commented on modules/tensor_mechanics/src/kernels/PoroMechanicsCoupling.C:
```diff
@@ -0,0 +1,53 @@
+{
+  if (_var.number() != _porepressure_var_num)
+    return 0.0;
+  return -_coefficient.value(_t, _q_point[_qp])*_phi[_j][_qp]*_grad_test[_i][_qp](_component);
```
This line seems wrong to me. I assume that this kernel subtracts the pore pressure from the stress divergence equation. The stress divergence equation is used to solve for the components of the displacement and since the residual isn't a function of the displacements, shouldn't the jacobian be zero?

WilkAndy replied:
Have a look at the jacobian test poro01.i .  I'm probably being over cautious here.  Yes, @tonkmr , usually this jacobian would be zero for the reason you state, BUT, i put this nonzero jacobian here just in case someone uses this kernel with _var = _porepressure_var  , ie, not for poromechanics, but something else.  (EG, the DE might be \nabla^2 u + \nabla_x u = 0).  Perhaps i'm being silly, but what's wrong with being cautious?

dschwen replied:
Uhm, right, @WilkAndy , what does happen if you couple to the same variable you that the kernel is acting on. Would MOOSE just call the on-digonal Jacobian method?

WilkAndy replied:
oh crap, thanks @dschwen , now i FINALY understand what @friedmud was confused about!  i'll alter the code next week.  This comes from writing half-a-line a day instead of making a concentrated effort.

a

tonkmr commented:
@WilkAndy adding an extra term to the displacement residuals to subtract off the pore pressure would be identical to using the extra_stress hook I have added to ComputeStressBase. In the material you could make the extra_stress be a function of the pore pressure variable. The one thing you do gain here is easier access to adding off diagonal terms to the Jacobian. If this is the preferred approach, then we should probably get rid of the extra_stress property.

WilkAndy commented:
Hey guys, I'm rather distracted by kids at the moment, but i had a play and can't really think of a cleaner way of writing the Jacobian code.  Is it just too obscure?  Can you think of a clearer presentation?

WilkAndy commented:
@tonkmr - let me have another look at this.  I recall that it added the extra_stress after the return-map, which i don't think is what i want.  But let me have a look - i'm sorry for such slow progress.

tonkmr commented:
@WilkAndy I'm fine with you doing it either way, but splitting it out in a new kernel as you have done here is outside of the return map as well. Even though it is outside of the nested solve at each quad point, it is included in the stress divergence residual so it will be properly accommodated for during the PETSc solve.

WilkAndy commented:
Hi @tonkmr .  In many ways i like the extra_stress idea because it seems really flexible.  On the other hand, there are two disadvantages: as you say, it making the Jacobian calculations probably more obscure since presumably we'd have to put them somewhere less obvious than just in a single Kernel; also it means that the meaning of 'stress' is different at different places in the code.  Eg, in my case, 'stress' would mean 'effective stress' in the plastic-return map, but it would mean 'total stress' in the Kernel calculations.
I vote we keep the PoroMechanicsCoupling kernel, but we don't remove 'extra_stress', as it might be useful later.

WilkAndy commented:
I'm really happy this has generated a lot of interest - thank you for your time and effort everyone.

I AM a bit stuck with how to proceed with the Jacobian calculations, as I'm quite happy with the code at present.  It would be great if someone would like to re-write the code, but i would like the tests/jacobian/poro01.i to pass, and not just a restricted version of it.

WilkAndy committed:
Material and Kernel added that model the fluid side of poromechanics
with a single-phase, fully-saturated fluid that has constant bulk modulus.

Fixes #4792

WilkAndy committed:
Tests of poromechanics for the single-phase, fully-saturated case.

Undrained Oedometer test
Porepressure generation test
Unconfined porepressure generation test
Unconsolidated undrained test
Terzaghi consolidation test

Fixes #4792

moosebuild commented:
Results of testing f2f7605085d96da68c789fbef1f0a481d58f28c8 using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14023

WilkAndy commented:
Hi everyone,

I'd really like @pou036 , or someone else, to look over the Material and Kernel i just added.  It passes quite a nontrivial test (terzaghi consolidation), but i'm super tired at the moment, and this was a rush job, so the code is probably suboptimal.  Could you give feedback, @pou036 ?

a

moosebuild commented:
Results of testing f2f7605085d96da68c789fbef1f0a481d58f28c8 using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14026

moosebuild commented:
Results of testing f2f7605085d96da68c789fbef1f0a481d58f28c8 using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14024

moosebuild commented:
Results of testing f2f7605085d96da68c789fbef1f0a481d58f28c8 using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14025

WilkAndy commented:
Heheheh, there's no need to look at this over Easter!  I'm working so slowly it'll be some weeks before this PR is finalised.

moosebuild commented:
Results of testing 1a0965c7d1f9dda760e6f9849acd98249da7568b using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14250

moosebuild commented:
Results of testing 1a0965c7d1f9dda760e6f9849acd98249da7568b using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14253

moosebuild commented:
Results of testing 1a0965c7d1f9dda760e6f9849acd98249da7568b using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14251

moosebuild commented:
Results of testing 1a0965c7d1f9dda760e6f9849acd98249da7568b using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14252

moosebuild commented:
Results of testing ae0d9f3f52f1c8576e86d65f70a291ff373f9656 using moose_PR_pre_check recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14595

moosebuild commented:
Results of testing ae0d9f3f52f1c8576e86d65f70a291ff373f9656 using moose_PR_pre_check recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14595

moosebuild commented:
Results of testing 2c245e0b0b87145613d85aca034cb78cf747fb3e using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14599

moosebuild commented:
Results of testing 2c245e0b0b87145613d85aca034cb78cf747fb3e using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14602

moosebuild commented:
Results of testing 2c245e0b0b87145613d85aca034cb78cf747fb3e using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14600

moosebuild commented:
Results of testing 2c245e0b0b87145613d85aca034cb78cf747fb3e using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14601

WilkAndy committed:
PoroMechanicsCoupling now has the Biot coefficient coming from a Material property, rather than an independent parameter.  This makes input files less error prone, smaller, and clearer.

Fixes #4792

moosebuild commented:
Results of testing 1fb4b165fac622c46598ce5f673486c93f95c840 using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14671

moosebuild commented:
Results of testing 1fb4b165fac622c46598ce5f673486c93f95c840 using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14672

moosebuild commented:
Results of testing 1fb4b165fac622c46598ce5f673486c93f95c840 using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14674

moosebuild commented:
Results of testing 1fb4b165fac622c46598ce5f673486c93f95c840 using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14673

WilkAndy committed:
Mandel compression-of-a-poroelastic-sample test added.
Poroelastic response of a borehole test added.

These tests, and the others, have been described on the MOOSE Framework wiki
http://mooseframework.org/wiki/PhysicsModules/TensorMechanics/TensorMechanicsBasics/PoroMechanics/

Fixes #4792

moosebuild commented:
Results of testing 9936daf788652da8451412f8b1e1f1ac75aef2d4 using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14924

moosebuild commented:
Results of testing 9936daf788652da8451412f8b1e1f1ac75aef2d4 using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14927

moosebuild commented:
Results of testing 9936daf788652da8451412f8b1e1f1ac75aef2d4 using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14925

moosebuild commented:
Results of testing 9936daf788652da8451412f8b1e1f1ac75aef2d4 using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/14926

moosebuild commented:
Results of testing 9e4db2a7f1ab283ef09687bd029301f0f3d2f87e using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15265

WilkAndy commented:
Finally I'm happy with the number of tests, for the fully-saturated single-phase case anyway.  It would be extremely unlikely if the implementation was incorrect after passing these fairly nontrivial tests.

moosebuild commented:
Results of testing 9e4db2a7f1ab283ef09687bd029301f0f3d2f87e using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15268

moosebuild commented:
Results of testing 9e4db2a7f1ab283ef09687bd029301f0f3d2f87e using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15266

moosebuild commented:
Results of testing 9e4db2a7f1ab283ef09687bd029301f0f3d2f87e using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15267

moosebuild commented:
Results of testing 259fc7f38b34a6e62fe6ec56ec7f459951e7432a using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15357

WilkAndy commented:
If anyone's keen, i've spent some time documenting the equations, kernels, tests, etc 

http://mooseframework.org/wiki/PhysicsModules/TensorMechanics/TensorMechanicsBasics/PoroMechanics/

moosebuild commented:
Results of testing 259fc7f38b34a6e62fe6ec56ec7f459951e7432a using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15360

moosebuild commented:
Results of testing 259fc7f38b34a6e62fe6ec56ec7f459951e7432a using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15358

moosebuild commented:
Results of testing 259fc7f38b34a6e62fe6ec56ec7f459951e7432a using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15359

WilkAndy committed:
PoroMechanicsAction added.

Fixes #4792

WilkAndy committed:
Tests of the PoroMechanics Action.  Also, re-wrote the borehole tests because they used an outdated version of the MultiPlasticity stuff.

Fixes #4792

moosebuild commented:
Results of testing 9f2b159af9f9d5b0bf01a794ecb79eb3d6e524ff using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/16721

moosebuild commented:
Results of testing 9f2b159af9f9d5b0bf01a794ecb79eb3d6e524ff using moose_PR_app_tests recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/16724

moosebuild commented:
Results of testing 9f2b159af9f9d5b0bf01a794ecb79eb3d6e524ff using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/16722

moosebuild commented:
Results of testing 9f2b159af9f9d5b0bf01a794ecb79eb3d6e524ff using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/16723

WilkAndy commented:
Hi everyone,

I don't think the test should fail now.  If everything's OK, could you please merge this?  I actually haven't addressed the multiphase situation, but i'm not sure: (1) exactly the best way of doing it; (2) whether i'll have time to do it for quite a while.  The present code is perfectly fine without the multiphase stuff, and people are already using it, so it'd be convenient to have it merged.

a

moosebuild commented:
Results of testing 5b8abd563c46b11ec5497f7b8e90a528980f2aed using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18453

moosebuild commented:
Results of testing 5b8abd563c46b11ec5497f7b8e90a528980f2aed using moose_PR_app_tests recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18456

WilkAndy commented:
Why does the app_tests recipe always fail?  What have i done wrong?

a

dschwen commented:
I wonder if I may be responsible. It is probably testing against a new
bison version that uses an integration branch and will not work with you
moose version.

On Mon, Jun 22, 2015, 9:12 PM Andy Wilkins <notifications@github.com> wrote:

> Why does the app_tests recipe always fail? What have i done wrong?
>
> a
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/4841#issuecomment-114338986>.
>

WilkAndy commented:
OK... so how do i fix it?

a

moosebuild commented:
Results of testing 5b8abd563c46b11ec5497f7b8e90a528980f2aed using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18454

moosebuild commented:
Results of testing 5b8abd563c46b11ec5497f7b8e90a528980f2aed using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18455

WilkAndy commented:
Hey @dschwen - do you want to suggest ideas for getting this PR to pass the tests?

dschwen commented:
If you rebased my changes are not responsible anymore. Try that.

On Thu, Jun 25, 2015, 8:21 PM Andy Wilkins <notifications@github.com> wrote:

> Hey @dschwen <https://github.com/dschwen> - do you want to suggest ideas
> for getting this PR to pass the tests?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/4841#issuecomment-115464968>.
>

moosebuild commented:
Results of testing e8725b690dcae527ee7972c616570165684d47fa using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18688

moosebuild commented:
Results of testing e8725b690dcae527ee7972c616570165684d47fa using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18689

moosebuild commented:
Results of testing e8725b690dcae527ee7972c616570165684d47fa using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18691

WilkAndy commented:
cool!  Thanks @dschwen ,

a

moosebuild commented:
Results of testing e8725b690dcae527ee7972c616570165684d47fa using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18690

tonkmr commented on modules/combined/tests/poro_mechanics/borehole_lowres.i:
```diff
@@ -0,0 +1,733 @@
+# Poroelastic response of a borehole.
+#
+# LOWRES VERSION: this version does not give perfect agreement with the analytical solution
```
Both the low res and high res versions say that they don't give perfect agreement with the analytical solution. Are one of these comments wrong? Does one of them agree with the analytical solution?

WilkAndy replied:
oops, thanks for noting that @tonkmr.  The highres version does agree with the analytical solution, as you would hope!  I'll do a change in a few hours.

a

tonkmr commented:
Let me know when this is ready to be merged.

moosebuild commented:
Results of testing 35a1c4b2e0eaff0831d7d5079d1cbc2198789393 using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18692

moosebuild commented:
Results of testing 35a1c4b2e0eaff0831d7d5079d1cbc2198789393 using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18695

moosebuild commented:
Results of testing 35a1c4b2e0eaff0831d7d5079d1cbc2198789393 using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18693

moosebuild commented:
Results of testing 35a1c4b2e0eaff0831d7d5079d1cbc2198789393 using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/18694

WilkAndy commented:
Hi @tonkmr , it looks like everything is ready to merge.  Only took me 3 months to complete this !

a

WilkAndy commented:
Wow, thanks @tonkmr , this was a long-running project - nice to have it finished,

a
